src/main/scala/nl.knaw.dans.easy.deposit/StateManager.scala

```diff
@@ -0,0 +1,110 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package nl.knaw.dans.easy.deposit
+
+import java.util.UUID
+
+import better.files.File
+import nl.knaw.dans.easy.deposit.Errors.{ IllegalStateTransitionException, InvalidPropertyException, PropertyNotFoundException }
+import nl.knaw.dans.easy.deposit.docs.StateInfo
+import nl.knaw.dans.easy.deposit.docs.StateInfo.State
+import nl.knaw.dans.easy.deposit.docs.StateInfo.State.State
+import org.apache.commons.configuration.PropertiesConfiguration
+
+import scala.util.{ Failure, Success, Try }
+
+case class StateManager(depositDir: File, submitBase: File) {
+  private val stateDescriptionKey = "state.description"
+  private val stateLabelKey = "state.label"
+  private val bagIdKey = "bag-store.bag-id"
+  val draftProps = new PropertiesConfiguration(
+    (depositDir / "deposit.properties").toJava
+  )
+  private lazy val submittedProps = new PropertiesConfiguration(
+    (submitBase / getProp(bagIdKey) / "deposit.properties").toJava
+  )
+
+  def getStateInfo: Try[StateInfo] = Try {
+    val draftState = getStateLabel()
+    draftState match {
+      case State.submitted | State.inProgress =>
+        val newState: StateInfo = getProp(stateLabelKey, submittedProps) match {
+          case "SUBMITTED" => StateInfo(State.submitted, getStateDescription())
+          case "REJECTED" => StateInfo(State.rejected, getStateDescription(submittedProps))
+          case "FEDORA_ARCHIVED" => StateInfo(State.archived, "The dataset is published in https://easy.dans.knaw.nl/ui")
+          case "IN_REVIEW" | "FAILED" => StateInfo(State.inProgress, "The dataset is visible for you under your datasets in https://easy.dans.knaw.nl/ui")
+          case str: String => throw InvalidPropertyException(stateLabelKey, str, submittedProps)
+        }
+        saveNewState(newState)
+      case State.draft | State.rejected | State.archived =>
+        StateInfo(draftState, getStateDescription())
+    }
+  }
+
+  def canChangeState(newStateInfo: StateInfo): Try[Unit] = getStateInfo.flatMap { oldStateInfo =>
+    val oldState = oldStateInfo.state
+    val newState = newStateInfo.state
+    (oldState, newState) match {
+      case (State.draft, State.submitted) |
+           (State.rejected, State.draft) => Success(())
+      case _ => Failure(IllegalStateTransitionException(oldState, newState))
+    }
+  }
+
+  /** @return bag-store.bag-id in case the new state is submitted */
+  def changeState(newStateInfo: StateInfo): Try[UUID] = getStateInfo.flatMap { old =>
```

Comment:
don't return `Success(null)`, but use `Option`; then the return type would be `Try[Option[UUID]]`